mlir/lib/Quantum/Transforms/ConversionPatterns.cpp

```diff
@@ -66,6 +66,75 @@ Value getGlobalString(Location loc, OpBuilder &rewriter, StringRef key, StringRe
         rewriter.create<LLVM::AddressOfOp>(loc, glb), ArrayRef<Value>({idx, idx}));
 }
 
+Value getModifiersPtr(Location loc, OpBuilder &rewriter, TypeConverter *conv, bool adjoint,
+                      ValueRange controlledQubits, ValueRange controlledValues)
+{
+    MLIRContext *ctx = rewriter.getContext();
+
+    auto boolType = IntegerType::get(ctx, 1);
+    auto sizeType = IntegerType::get(ctx, 64);
+    auto sizePtrType = LLVM::LLVMPointerType::get(IntegerType::get(ctx, 64));
+    auto qubitType = conv->convertType(QubitType::get(ctx));
+    auto qubitPtrType = LLVM::LLVMPointerType::get(qubitType);
+    auto qubitPtrPtrType = LLVM::LLVMPointerType::get(qubitPtrType);
+    auto boolPtrType = LLVM::LLVMPointerType::get(boolType);
+    auto boolPtrPtrType = LLVM::LLVMPointerType::get(boolPtrType);
+    auto structType = LLVM::LLVMStructType::getLiteral(
+        ctx, {boolType, sizeType, qubitPtrType, boolPtrType}, true);
+    auto modifiersPtrType = LLVM::LLVMPointerType::get(structType);
+
+    Value c0 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(0));
+    Value c1 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(1));
+    Value nullPtr = rewriter.create<LLVM::IntToPtrOp>(loc, qubitPtrType, c0);
+
+    if (!adjoint && controlledQubits.empty() && controlledValues.empty()) {
+        return rewriter.create<LLVM::IntToPtrOp>(loc, modifiersPtrType, c0).getResult();
+    }
+
+    auto adjointVal = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getBoolAttr(adjoint));
+    auto numControlledVal = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(0));
+
+    auto modifiersPtr = rewriter.create<LLVM::AllocaOp>(loc, modifiersPtrType, c1).getResult();
+    auto adjointPtr = rewriter.create<LLVM::GEPOp>(loc, boolPtrType, modifiersPtr,
+                                                   llvm::ArrayRef<LLVM::GEPArg>{0, 0});
+    auto numControlledPtr = rewriter.create<LLVM::GEPOp>(loc, sizePtrType, modifiersPtr,
+                                                         llvm::ArrayRef<LLVM::GEPArg>{0, 1});
+    auto controlledWiresPtr = rewriter.create<LLVM::GEPOp>(loc, qubitPtrPtrType, modifiersPtr,
+                                                           llvm::ArrayRef<LLVM::GEPArg>{0, 2});
+    auto controlledValuesPtr = rewriter.create<LLVM::GEPOp>(loc, boolPtrPtrType, modifiersPtr,
+                                                            llvm::ArrayRef<LLVM::GEPArg>{0, 3});
+
+    Value ctrlPtr = nullPtr;
+    Value valuePtr = nullPtr;
+    if (controlledQubits.size() > 0) {
+        Value c = rewriter.create<LLVM::ConstantOp>(
+            loc, rewriter.getI64IntegerAttr(controlledQubits.size()));
```

Comment:
Can't we reuse `numControlledVal` from above?

Reply:
Done